### PR TITLE
Cache entity metadata properly

### LINKED ISSUE #1053: Memory Leak - JobsEntityDamagePlayer
***Description of issue:**

It Seems the entityDamageByPlayer metadata is not being cleared anywhere in the code, and it's never cleared by the GC, maybe add something to clear the metadata on mob kill, or a timer?

---

imagen

---

**Server Type (Spigot/Paperspigot/etc):**  paper

**Server Version (using `/ver`):**  This server is running Paper version git-Paper-399 (MC: 1.16.4) (Implementing API version 1.16.4-R0.1-SNAPSHOT)

### PR DESCRIPTION
**Pull request by MrMacor**
This fixes #1053 by replacing the previous entity metadata system with a proper caching system for tracking how much damage players have inflicted on mobs, how long since a cow has been milked, and if an end crystal's explosion was caused by a player. 

I'm marking this as a draft because I might implement a similar system for the block ownership system, and I'm not sure if the current changes are breaking (some help testing would be appreciated).

### CODE REVIEW

File: src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java [hunk 1]
```diff
@@ -210,6 +218,7 @@ public void onCowMilking(PlayerInteractEntityEvent event) {
 	    return;
 
 	Entity cow = event.getRightClicked();
+	UUID cowUUID = cow.getUniqueId();
```
Comment (montlikadani) [inline src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java:221]: Move this below before the `CowMilkingTimer` condition.

File: src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java [hunk 2]
```diff
@@ -1062,6 +1070,7 @@ public void onEntityDamageByPlayer(EntityDamageEvent event) {
 	    return;
 
 	Entity ent = event.getEntity();
+	UUID entUUID = ent.getUniqueId();
```
Comment (montlikadani) [inline src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java:1073]: Move this below before the `MonsterDamageUse` condition.

File: src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java [hunk 3]
```diff
@@ -239,8 +248,8 @@ public void onCowMilking(PlayerInteractEntityEvent event) {
 	}
 
 	if (Jobs.getGCManager().CowMilkingTimer > 0) {
-	    if (cow.hasMetadata(cowMetadata)) {
-		long time = cow.getMetadata(cowMetadata).get(0).asLong();
+	    if (cowMilkingTimer.getIfPresent(cowUUID) != null) {
```
Comment (montlikadani) [inline src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java:251]: Declare and cache `getIfPresent` to 'Long' to make sure the process is fast enough, as `Map#get` is a bit slow.

File: src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java [hunk 4]
```diff
@@ -94,19 +87,34 @@
...
-import java.util.UUID;
+import java.util.concurrent.TimeUnit;
 
 public class JobsPaymentListener implements Listener {
 
     private Jobs plugin;
-
-    private final String blockMetadata = "BlockOwner", cowMetadata = "CowTimer", entityDamageByPlayer = "JobsEntityDamagePlayer";
+    private final Cache<UUID, Double> damageDealtByPlayers = CacheBuilder.newBuilder()
+		    .expireAfterWrite(10, TimeUnit.MINUTES)
```
Comment (montlikadani) [inline src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java:98]: I think 5 minutes or less would be enough for the expiration time.

File: src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java [hunk 5]
```diff
@@ -239,8 +248,8 @@ public void onCowMilking(PlayerInteractEntityEvent event) {
 	}
 
 	if (Jobs.getGCManager().CowMilkingTimer > 0) {
-	    if (cow.hasMetadata(cowMetadata)) {
-		long time = cow.getMetadata(cowMetadata).get(0).asLong();
+	    if (cowMilkingTimer.getIfPresent(cowUUID) != null) {
+		long time = cowMilkingTimer.getIfPresent(cowUUID);
```
Comment (montlikadani) [inline src/main/java/com/gamingmesh/jobs/listeners/JobsPaymentListener.java:252]: Declare and cache `getIfPresent` to `Long` to make sure the process is fast enough, as `Map#get` is a bit slow.
```suggestion
	    Long time = cowMilkingTimer.getIfPresent(cowUUID);
	    if (time != null) {
```

**Comment by montlikadani** [comment]
> I'm marking this as a draft because I might implement a similar system for the block ownership system, and I'm not sure if the current changes are breaking (some help testing would be appreciated).

There is no need to rework the blockOwnerShip metadata if a soft referenced metadata is handled correctly, without user reporting and based on tests, there is no such observation that there is a memory leak in this case.

**Comment by MrMacor** [comment]
> There is no need to rework the blockOwnerShip metadata if a soft referenced metadata is handled correctly, without user reporting and based on tests, there is no such observation that there is a memory leak in this case.

Yup, you're right. I've also committed a change that moves some code around to make sure that `jobsMobSpawner` is always removed from entities when they die. The only issue I can think of now are when mobs are spawned in by a spawner, the `jobsMobSpawner` meta is added, and then it may not get removed if the entity is never killed (there's no way to remove the metadata when the entity despawns as Spigot doesn't have Paper's EntityRemoveFromWorldEvent). Do you think we should cache spawner mob UUIDs with a similar setup to the damage cache?

**Comment by montlikadani** [comment]
> The only issue I can think of now are when mobs are spawned in by a spawner, the `jobsMobSpawner` meta is added, and then it may not get removed if the entity is never killed (there's no way to remove the metadata when the entity despawns as Spigot doesn't have Paper's EntityRemoveFromWorldEvent). Do you think we should cache spawner mob UUIDs with a similar setup to the damage cache?

Instead of Paper's event we could use and listen to ChunkUnloadEvent or some related event that will remove the mobs metadata, obviously if it is called. I'm unsure that this event ever called, I never used.

**Comment by MrMacor** [comment]
> Instead of Paper's event we could use and listen to ChunkUnloadEvent or some related event that will remove the mobs metadata, obviously if it is called. I'm unsure that this event ever called, I never used.

Good idea, I've implemented this now.

**Comment by montlikadani** [comment]
Hopefully works...

**Pull request merged**